docs: Add integrations/Packs doc for ngrok Kubernetes Ingress Controller (#1612)
spectrocloud/librarium: docs/docs-content/integrations/ngrok.md

---

The ngrok [Ingress Controller for Kubernetes](https://github.com/ngrok/kubernetes-ingress-controller) adds public and
secure ingress traffic to Kubernetes applications. This open-source [Ingress
Controller](https://kubernetes.io/docs/concepts/services-networking/ingress-controllers) works with Palette to provide
ingress to your applications, APIs, or other resources while also offloading network ingress and middleware execution to
ngrok's platform.

# Versions Supported

- 0.9.0

## Prerequisites

- An active [ngrok account](https://ngrok.com/signup).
  - An ngrok authentication token. You can find your token in the dashboard. Visit the [**Your Authtoken**](https://dashboard.ngrok.com/get-started/your-authtoken) section to review your access token.
  - An ngrok API key. You can generate an API key from the ngrok dashboard. Visit the [**API** section](https://dashboard.ngrok.com/api) of the dashboard to review existing keys.
- A static subdomain. You can obtain a static subdomain by navigating to the [**Domains**
section](https://dashboard.ngrok.com/cloud-edge/domains) of the ngrok dashboard and clicking on **Create Domain** or **New Domain**.

## Parameters

> To deploy the ngrok Ingress Controller, you need to set, at minimum, the following parameters your preset or active pack profile.

Reviewer: ```suggestion
To deploy the ngrok Ingress Controller, you need to set, at minimum, the following parameters in the pack's YAML.
```

---
sidebar_label: "ngrok"
title: "ngrok"
description: "Learn about using ngrok Kubernetes Ingress to access applications and services in Spectro Cloud."
hide_table_of_contents: true
type: "integration"
category: ["ingress", "kubernetes"]
sidebar_class_name: "hide-from-sidebar"
logoUrl: ""
tags: ["packs", "ngrok", "network", "kubernetes"]
---

> The ngrok [Ingress Controller for Kubernetes](https://github.com/ngrok/kubernetes-ingress-controller) is the official

Reviewer: @joelhans the word "official" raises a minor concern. Can we change this sentence so it's more apparent that it's ngrok's official ingress controller? Right now, it could potentially be perceived that this is our official ingress controller or the Kubernetes ecosystem's official controller.

controller for adding public and secure ingress traffic to your k8s services. This open source [Ingress

> Controller](https://kubernetes.io/docs/concepts/services-networking/ingress-controllers) works with Spectro Cloud to

Reviewer: ```suggestion
Controller](https://kubernetes.io/docs/concepts/services-networking/ingress-controllers) works with Palette to
```

provide ingress to your applications, APIs, or other services while also offloading network ingress and middleware
execution to ngrok's platform.

# Versions Supported

- 0.9.0

## Prerequisites

- An active [ngrok account](https://ngrok.com/signup).

>   - Your ngrok authentication token (`AUTHTOKEN`), which you can find under under [**Your
>    Authtoken**](https://dashboard.ngrok.com/get-started/your-authtoken) in the ngrok dashboard.

Reviewer: ```suggestion
  - An ngrok authentication token.  You can find your token in the dashboard. Visit the [**Your
   Authtoken**](https://dashboard.ngrok.com/get-started/your-authtoken) section to review our access token.
```
Updating this bullet to fix indentation and match our voice.

>   - An API key (`API_KEY`), which you can generate in the [**API** section](https://dashboard.ngrok.com/api) of the
>     ngrok dashboard.

Reviewer: ```suggestion
  - An ngrok API key. You can generate an API key from the ngrok dashboard. Visit the [**API** section](https://dashboard.ngrok.com/api) of the dashboard to create and review existing key.
```
Touched up the sentence to better match our style.

> - A static subdomain (`NGROK_DOMAIN`), which you can obtain by navigating to the [**Domains**
>    section](https://dashboard.ngrok.com/cloud-edge/domains) of the ngrok dashboard and clicking **Create Domain** or
>    **New Domain**.

Reviewer: ```suggestion
- A static subdomain. You can obtain a static subdomain by navigating to the [**Domains**
   section](https://dashboard.ngrok.com/cloud-edge/domains) of the ngrok dashboard and clicking on **Create Domain** or **New Domain**.
```

> ## Parameters

Reviewer: @joelhans for this section, please create a markdown table containing all the required parameters; API_KEY, AUTHTOKEN, and DOMAIN.
This table should be placed at the top of the parameters section. The table requires the two following columns; Parameter, Description

This section is often associated with the YAML of the pack itself. Are there other common parameters worth highlighting an placing in the markdown table? you don't have add every single parameter available in the YAML, just the ones worth being aware off. You can link to these resources that you all have created:
 - [Common Helm Overrides](https://github.com/ngrok/kubernetes-ingress-controller/blob/main/docs/deployment-guide/common-helm-k8s-overrides.md)
 -  [ngrok Kubernetes Ingress Controller Documentation](https://github.com/ngrok/kubernetes-ingress-controller/blob/main/docs/README.md#ngrok-kubernetes-ingress-controller-documentation)

![CleanShot 2023-09-27 at 07 31 00](https://github.com/spectrocloud/librarium/assets/29551334/9d7d1bf4-ad6a-4942-94f4-bbc9c38a8051)

Author: Thanks for the tip here. I have added the *required* parameters into this table and then linked to resources where users could find more advanced configurations, like fan-outs, which mostly work via adding additional `path` parameters.

> To deploy properly, the ngrok Kubernetes Ingress Controller requires your ngrok authentication `API_KEY` and
> `AUTHTOKEN`, which you can add to your preset or running profile.

Reviewer: ```suggestion
To properly deploy the ngrok Kubernetes Ingress Controller, you need to add your ngrok authentication `API_KEY` and `AUTHTOKEN` to your preset or active profile.
```
Touched up the sentence to make it easier to read.

```yaml
charts:  
  kubernetes-ingress-controller:
    credentials:
      apiKey: API_KEY
      authtoken: AUTHTOKEN
```

> :::tip
> 
> Free ngrok accounts can have only one ngrok agent running at a time. If you have a free ngrok account, change
> the number of replicas to `1`, like the following:

Reviewer: ```suggestion
:::caution

If you have a free ngrok account, you can only have one ngrok agent active at a time. This means that you will need to set the number of replicas to `1` to ensure that your ngrok agent operates properly.
```

```yaml
charts:
  kubernetes-ingress-controller:
    replicaCount: 1
```

:::

> ## Usage

Reviewer: @joelhans I like how you included an example but I think we need to explain a bit more what is happening here. This section should focus on explaining the `ingress` section of the YAML and why the current values are used.

 Think of this section as helping the reader understand how to use the ngrok controller pack. Don't duplicate information, link to existing resources that you all have created, but do go into some details of what is happening here so the reader can understand.

> The following `.yaml` file demonstrates an example deployment using the [2048](https://github.com/gabrielecirulli/2048)

Reviewer: ```suggestion
The following YAML content demonstrates a deployment using the [2048](https://github.com/gabrielecirulli/2048)
```

you created in the previous step. It should look something like `one-two-three.ngrok-free.app`.

:::

```yaml
apiVersion: v1
kind: Service
metadata:
  name: game-2048
  namespace: ngrok-ingress-controller
spec:
  ports:
    - name: http
      port: 80
      targetPort: 80
  selector:
    app: game-2048
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: game-2048
  namespace: ngrok-ingress-controller
spec:
  replicas: 1
  selector:
    matchLabels:
      app: game-2048
  template:
    metadata:
      labels:
        app: game-2048
    spec:
      containers:
        - name: backend
          image: alexwhen/docker-2048
          ports:
            - name: http
              containerPort: 80
---
# ngrok Ingress Controller Configuration
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: game-2048-ingress
  namespace: ngrok-ingress-controller
spec:
  ingressClassName: ngrok
  rules:
    - host: NGROK_DOMAIN
      http:
        paths:
          - path: /
            pathType: Prefix
            backend:
              service:
                name: game-2048
                port:
                  number: 80
```

> 
> Apply the `2048.yaml` manifest to your cluster in Palette.
> 
> ```bash
> kubectl apply -f 2048.yaml
> ```
> 
> Access your 2048 demo app by navigating to the your `NGROK_DOMAIN`, e.g. `https://one-two-three.ngrok-free.app`. ngrok's
> edge and your Ingress Controller will route traffic to your app from any device or external network.

Reviewer: ```suggestion
```
We don't need these steps, because this is not intended to be a tutorial. You can link to your actual [tutorial](https://ngrok.com/docs/using-ngrok-with/k8s/) if the user wants to learn more through a hands-on example. The important thing here is for the user to understand that they need to add the pack to the cluster profile, configure the pack YAML, and then for their application, they need to create a ingress service definition. 

The question I would ask as a newer users is. How do I create the ingress service for my application, assuming the pack is in the cluster profile. That's the piece we are missing. And for that, the answer is to create a [manifest layer.](https://docs.spectrocloud.com/cluster-profiles/task-define-profile?type=Add+Manifest)

Like this:

https://github.com/spectrocloud/librarium/assets/29551334/0d7454ea-b919-4463-8f86-c18628c8271c

This is really want the usage section should be about.

> 

Reviewer: We need to add a Terrafrom section. Like this:

```suggestion
## Terraform

You can reference the ngrok Ingress Controller pack in Terraform with a data resource.

data "spectrocloud_registry" "public_registry" {
  name = "Public Repo"
}

data "spectrocloud_pack_simple" "ngrok-ingress" {
  name    = "ngrok-ingress-controller"
  version = "0.9.0"
  type = "operator-instance"
  registry_uid = data.spectrocloud_registry.public_registry.id
}
```

Author: I admittedly have no experience in Terraform, so if we need information beyond this, I will need some more assistance. But I've added (not yet commited+pushed) this to the doc.

## References

> 
> - [Ingress Controller for Kubernetes on GitHub](https://github.com/ngrok/kubernetes-ingress-controller)
> - [ngrok documentation](https://ngrok.com/docs/)
> - [Get started with the ngrok Ingress Controller for Kubernetes](https://ngrok.com/docs/using-ngrok-with/k8s/#get-started-with-the-ngrok-ingress-controller-for-kubernetes)

Reviewer: ```suggestion

- [Ingress Controller for Kubernetes on GitHub](https://github.com/ngrok/kubernetes-ingress-controller)
- [ngrok documentation](https://ngrok.com/docs/)
- [Get started with the ngrok Ingress Controller for Kubernetes](https://ngrok.com/docs/using-ngrok-with/k8s/#get-started-with-the-ngrok-ingress-controller-for-kubernetes)
- [ngrok Pack GitHub](https://github.com/spectrocloud/pack-central/tree/main/packs/ngrok-ingress-controller-0.9.0)
- [ngrok Ingress Controller Helm Documentation](https://github.com/ngrok/kubernetes-ingress-controller/tree/main/docs)
```
Added some more resources I think advanced users would like to know.
